biolearn/data_library.py

```diff
@@ -405,6 +416,173 @@ def _metadata_load_list(self):
+                        cols_data.append(extract_informal_age(characteristics))
+                else:
+                    cols_data.append(extract_numeric(characteristics.get(key, "")))
+            else:
+                cols_data.append(characteristics.get(key))
+
+        return cols_data
+
+    def _convert_to_metadata_df(self, metadata_keys, samples):
+
+        data = {sample["acc"]: self._convert_characteristics_to_df_cols_data(
+            metadata_keys, sample) for sample in samples}
+
+        df = pd.DataFrame(data).transpose()
+        df = df.reset_index(drop=False)
+        columns = ["id"]
+        columns.extend(metadata_keys)
+        df.columns = columns
+        return df
+
+    def _get_matrix_file_size_from_url(self, url):
+        try:
+            response = requests.head(url)
+            response.raise_for_status()
+
+            content_length = response.headers.get('Content-Length')
+
+            if content_length is None:
+                print("Content-Length header is not present.")
+                return None
+
+            file_size = int(content_length)
+            return file_size
+
+        except requests.RequestException as e:
+            print(f"HTTP request failed: {e}")
+            return None
+
+    def _create_matrix(self, matrix_file):
+        matrix_file_path = cached_download(matrix_file)
+
+        row_num = self._get_matrix_table_row_num(matrix_file_path)
+
+        matrix_data = pd.read_table(
+            matrix_file_path, index_col=0, skiprows=row_num + 1
+        )
+        matrix_data = matrix_data.drop(
+            ["!series_matrix_table_end"], axis=0
+        )
+        matrix_data.index.name = "id"
+
+        return matrix_data
+
+    def _get_matrix_file_line_count(self, matrix_file):
+        def blocks(files, size=65536):
+            while True:
+                b = files.read(size)
+                if not b:
+                    break
+                yield b
+
+        with open(matrix_file, "r", encoding="utf-8", errors='ignore') as f:
+            return sum(bl.count("\n") for bl in blocks(f))
+
+    def _ungzip_file(self, matrix_file, output_file):
+        with gzip.open(matrix_file, 'rb') as f_in:
+            with open(output_file, 'wb') as f_out:
+                shutil.copyfileobj(f_in, f_out)
+
+    def _find_matrix_table_begin_line_num_backwards(self, matrix_file):
+        with FileReadBackwards(matrix_file) as frb:
+            line_num = 0
+            for l in frb:
+                line_num += 1
+                if "!series_matrix_table_begin" in l:
+                    break
+            return line_num
+
+    def _get_matrix_table_row_num(self, matrix_file):
+
+        output_file = f"{matrix_file}.txt"
+        self._ungzip_file(matrix_file, output_file)
+
+        total_lines = self._get_matrix_file_line_count(output_file)
+        line_num_bw = self._find_matrix_table_begin_line_num_backwards(output_file)
+
+        os.remove(output_file)
```

Comment:
It would probably be better if we didn't write to disk but it we do this should be in a finally block to ensure cleanup even if an exception is thrown.

Reply:
Okay, I will put it in the finally block to ensure it gets cleaned up.